src/lib.rs

```diff
@@ -1,5 +1,70 @@
 #![cfg_attr(not(feature = "use_std"), no_std)]
 
+#[derive(Debug)]
+pub struct CobsEncoder<'a> {
+    dest: &'a mut [u8],
+    dest_idx: usize,
+    code_idx: usize,
+    num_bt_sent: u8,
+}
+
+impl<'a> CobsEncoder<'a> {
+
+    /// Create a new streaming Cobs Encoder
+    pub fn new(out_buf: &'a mut [u8]) -> CobsEncoder<'a> {
+        CobsEncoder {
+            dest: out_buf,
+            dest_idx: 1,
+            code_idx: 0,
+            num_bt_sent: 1,
+        }
+    }
+
+    /// Push a slice of data to be encoded
+    pub fn push(&mut self, data: &[u8]) -> Result<(), ()> {
+        // TODO: could probably check if this would fit without
+        // iterating through all data
+        for x in data {
+            if *x == 0 {
+                *self.dest.get_mut(self.code_idx)
+                    .ok_or_else(|| ())? = self.num_bt_sent;
+
+                self.num_bt_sent = 1;
+                self.code_idx = self.dest_idx;
+                self.dest_idx += 1;
+            } else {
+                *self.dest.get_mut(self.dest_idx)
+                    .ok_or_else(|| ())? = *x;
+
+                self.num_bt_sent += 1;
+                self.dest_idx += 1;
+                if 0xFF == self.num_bt_sent {
+                    *self.dest.get_mut(self.code_idx)
+                        .ok_or_else(|| ())? = self.num_bt_sent;
+                    self.num_bt_sent = 1;
+                    self.code_idx = self.dest_idx;
+                    self.dest_idx += 1;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Complete encoding of the output message. Does NOT terminate
+    /// the message with the sentinel value
+    pub fn finalize(self) -> Result<usize, ()> {
```

Comment:
@awelkie do you think that the `finalize()` method should insert the final `0x00`? I don't think I did this, as it didn't match your current API, but IMO it makes sense to do.

Reply:
Yeah good question. I didn't insert `0x00` at the beginning or the end to let the user decide where to insert the sentinel. I figure there may be systems where the user would want to insert the sentinel at the beginning of the frame and not the end, or vice-versa, or maybe both, so I just avoided making that decision.

I'd prefer to keep it that way, but I also realize that in nearly every situation one would want to insert the sentinel at the end, so I'm open to debate.